### PR TITLE
Adding a DSPA ServiceMonitor template

### PR DESCRIPTION
**Pull request by DharmitD**
## Description
Adding a Service Monitor template to procure DSP Application metrics
Also added a CRD for ServiceMonitor because the unit tests run in a fake local K8s cluster which doesn't have the `monitoring.coreos.com/v1` APIVersion.

## How Has This Been Tested?

Ran `make deploy` with these changes
Created a DSP Application in a separate namespace. 
Checked if a ServiceMonitor named `ds-pipelines-samples` came up.
Checked that specifically Data Science Pipelines Application metrics were scraped.
Metrics were found to be scraped by Prometheus: 
image

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Review by harshad16 — COMMENTED** [review]
/lgtm
:+1:

File: config/crd/external/monitoring.coreos.com_servicemonitors.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: servicemonitors.monitoring.coreos.com
```
[the remark this answers is not in the document]
Comment (HumairAK) [reply config/crd/external/monitoring.coreos.com_servicemonitors.yaml:4]: this folder is mostly to inform our mock testing infra of crds not native to k8s fyi

[the remark this answers is not in the document]
Comment (DharmitD) [reply config/crd/external/monitoring.coreos.com_servicemonitors.yaml:4]: Yes the tests fail w/o it. The mock testing infra isn't able to identify the APIVersion maybe:

```
ERROR	Reconciler error	{"controller": "datasciencepipelinesapplication", "controllerGroup": "datasciencepipelinesapplications.opendatahub.io", "controllerKind": "DataSciencePipelinesApplication", "DataSciencePipelinesApplication": {"name":"testdsp0","namespace":"default"}, "namespace": "default", "name": "testdsp0", "reconcileID": "d7d879eb-4265-44e1-9c1d-1ca324ec501c", "error": "no matches for kind \"ServiceMonitor\" in version \"monitoring.coreos.com/v1\""}
```
With this CRD added, the tests pass.

**Comment by gregsheremeta** [comment]
this is marked as draft still. I can approve if you undraft it

**Comment by DharmitD** [comment]
> this is marked as draft still. I can approve if you undraft it

Had marked it as draft because the Service Monitor wasn't able to pull in the correct DSP Application metrics. Updated Network Policy, ServiceMonitor is able to pull in metrics now. PR is marked ready for review!

File: config/internal/common/policy.yaml.tmpl [hunk 2]
```diff
@@ -18,6 +18,13 @@ spec:
     # We only allow DSPA components to communicate
     # by bypassing oauth proxy, all external
     # traffic should go through oauth proxy
+    - from:
+        - namespaceSelector:
+            matchLabels:
+              name: openshift-user-workload-monitoring
```
Comment (harshad16) [inline config/internal/common/policy.yaml.tmpl:24]: please just add this part to from part below after line 28
that would be enough, separate from key would override rest of things.

**Review by harshad16 — CHANGES_REQUESTED** [review]
policy.yaml.tmpl format would cause issues.

File: config/internal/apiserver/monitor.yaml.tmpl [hunk 3]
```diff
@@ -0,0 +1,16 @@
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  labels:
+    app: ds-pipeline-{{.Name}}
```
Comment (harshad16) [inline config/internal/apiserver/monitor.yaml.tmpl:5]:
```suggestion
    app: ds-pipeline-servicemonitor-{{.Name}}
```
as we having different naming, lets follow the same pattern.

File: config/internal/common/policy.yaml.tmpl (hunk 2 above)
Comment (DharmitD) [reply config/internal/common/policy.yaml.tmpl:24]: Done, refactored based on this suggestion, pushed to an image and re-deployed. Metrics are scraped correctly:

Screenshot from 2023-03-23 15-52-52

**Review by harshad16 — COMMENTED** [review]
thanks, this looks good to me
just for good to have changes,
if we feel other about them , we can skip it.

**Comment by gregsheremeta** [comment]
one minor comment on Hashard's comment ... otherwise lgtm

File: config/internal/apiserver/monitor.yaml.tmpl (hunk 3 above)
Comment (harshad16) [reply config/internal/apiserver/monitor.yaml.tmpl:5]: I meant on labels.app itself, not on the other parts those looks good to me
anyhow this is a hard requirement, please ignore it.

Comment (gregsheremeta) [reply config/internal/apiserver/monitor.yaml.tmpl:5]: oh. Huh. I didn't notice that the labels.app is changing all over the code. I think I misunderstood what that was for.

Comment (DharmitD) [reply config/internal/apiserver/monitor.yaml.tmpl:5]: I kept the labels.app matching to what we have in the service and route:
https://github.com/opendatahub-io/data-science-pipelines-operator/blob/main/config/internal/apiserver/route.yaml.tmpl#L7
https://github.com/opendatahub-io/data-science-pipelines-operator/blob/main/config/internal/apiserver/service.yaml.tmpl#L9

The matchlabels in ServiceMonitor are also exactly the same as label.app in the service. My suggestion would be to keep the labels.app as is, consistent with the service and route. What do you think @harshad16 @gregsheremeta ?

Comment (gregsheremeta) [reply config/internal/apiserver/monitor.yaml.tmpl:5]: I hesitate to comment further because I don't know what it does :joy:

Comment (harshad16) [reply config/internal/apiserver/monitor.yaml.tmpl:5]: Technically `matchlabels` in `selector` section is the only important bit as per fetching metrics from apisever which absolutely correct written.
i was just suggesting on the labels, as labels in metadata are used for filtering.
anyhow this is good for me, no changes needed here
thanks @DharmitD

File: controllers/dspipeline_controller.go [hunk 4]
```diff
@@ -139,6 +139,7 @@ func (r *DSPAReconciler) DeleteResourceIfItExists(ctx context.Context, obj clien
 //+kubebuilder:rbac:groups=authentication.k8s.io,resources=tokenreviews,verbs=create
 //+kubebuilder:rbac:groups=image.openshift.io,resources=imagestreamtags,verbs=get
 //+kubebuilder:rbac:groups=core,resources=events,verbs=create;patch;list
+//+kubebuilder:rbac:groups=monitoring.coreos.com,resources=servicemonitors,verbs=get;list;watch;create;update;patch;delete
```
[the remark this answers is not in the document]
Comment (DharmitD) [reply controllers/dspipeline_controller.go:142]: Done, added same permissions to the ClusterRole.

**Review by harshad16 — COMMENTED** [review]
/lgtm

thanks :100:

**Review by harshad16 — APPROVED** [review]
/lgtm
/approve

great work :+1:

**Pull request merged**